### PR TITLE
#19 #20 Project set up

### PR DESCRIPTION
**Pull request by Cleop**
Sets the project up with heroku and sets up a new event endpoint that works with ngrok #19 #20 

Sets up CI

#1 Adds what/why/how to readme

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+elixir:
+  - 1.4
+addons:
+  postgresql: '9.4'
+env:
+  - MIX_ENV=test
+before_script:
+  - mix do ecto.create
+script:
+  - mix do deps.get, compile --warnings-as-errors, coveralls.json
```
Comment (nelsonic) [inline .travis.yml]: we might end up turning _off_ the `--warnings-as-errors` option. given the issues we're having in: https://github.com/dwyl/hits-elixir/issues/19

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+language: elixir
+elixir:
+  - 1.4
+addons:
+  postgresql: '9.4'
+env:
+  - MIX_ENV=test
+before_script:
+  - mix do ecto.create
```
Comment (nelsonic) [inline .travis.yml]: do we need a `mix ecto.migrate` here too?

File: README.md [hunk 3]
```diff
@@ -1,2 +1,164 @@
 # github-backup
 :back: :arrow_up: A script to backup your GitHub Account/Organisation so you can still work when (they/you are) offline.
+
+## Why?
+
+As a person who uses GitHub as their "_single-source-of-truth_"
```
Comment (nelsonic) [inline README.md]: this is _one_ of the _valid_ use-cases. 👍 
the _other_ is: 
We use "checklists" in GitHub issues as the "_acceptance criteria_".
to avoid ["scope creep"](https://en.wikipedia.org/wiki/Scope_creep) i.e. "_moving goalposts_",
(_where people add items to the checklist without accountability_) 
we need to _track_ the changes made to the issue description/comments.

File: README.md [hunk 4]
```diff
@@ -1,2 +1,164 @@
...
+I need a backup of GitHub issues
+so that I can work "offline".
+(_either when I have no network or GH is "down"_)
+
+## What?
+
+Backup GitHub issues.
+
++ a script that lets you login to GitHub using OAuth
```
Comment (nelsonic) [inline README.md]: it's no longer going to be a "script" more of a "_standalone app_" to view GitHub issue revision history.

File: README.md [hunk 5]
```diff
@@ -1,2 +1,164 @@
...
++ Use GitHub REST API to retrieve all issues
++ Store issues as "flat files" in subdirectories corresponding to the GitHub URL
+e.g: if the issue is `dwyl/github-reference/issues/15`
+store it as `dwyl/github-reference/issues/15.json`
++ It does not need to "scale" it just needs to work. read: http://paulgraham.com/ds.html
```
Comment (nelsonic) [inline README.md]: 👍

File: README.md [hunk 6]
```diff
@@ -1,2 +1,164 @@
...
+## Set up
+
+The project is hosted on heroku at: https://github-backup.herokuapp.com/
+
+To run it locally, you will need to:
+
+- Create a new Github application.
+- Run a phoenix server on your machine.
+
+You'll need to have installed [Elixir](https://elixir-lang.org/install.html), [Phoenix](https://hexdocs.pm/phoenix/installation.html), and [ngrok](https://ngrok.com/download) if you haven't already.
```
Comment (nelsonic) [inline README.md]: please separate these prerequisites into a bullet-point list and _explain_/_clarify_ why ngrok.com is needed.

File: assets/js/socket.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+//     end
+//
+// Finally, pass the token on connect as below. Or remove it
+// from connect if you don't care about authentication.
+
+socket.connect()
+
+// Now that you are connected, you can join channels with a topic:
+let channel = socket.channel("topic:subtopic", {})
+channel.join()
```
Comment (nelsonic) [inline assets/js/socket.js]: are we _using_ this?

File: mix.exs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  # Specifies your project dependencies.
+  #
+  # Type `mix help deps` for examples and options.
+  defp deps do
+    [
+      {:phoenix, "~> 1.3.0"},
+      {:phoenix_pubsub, "~> 1.0"},
+      {:phoenix_ecto, "~> 3.2"},
+      {:postgrex, ">= 0.0.0"},
```
Comment (nelsonic) [inline mix.exs]: I discovered today, after a _lot_ of debugging that the version of `postgrex` _cannot_ be `0.0.0` to _work_ on Heroku. see: https://github.com/dwyl/phoenix-chat-example/blob/036c895e0aff62c50d9f71ce98a8ee0106f16e3b/mix.exs#L42

File: .travis.yml [hunk 9]
```diff
@@ -0,0 +1,17 @@
+language: elixir
+elixir:
+  - 1.4
```
Comment (shouston3) [inline .travis.yml]: Why are we using `1.4`? `1.6` is out with some really cool features we could be using

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+addons:
+  postgresql: '9.4'
```
Comment (shouston3) [inline .travis.yml]: 9.6 was working fine for me on travis. (10 has issues though: https://github.com/travis-ci/travis-ci/issues/8537)

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+before_script:
+  - mix do ecto.create
+script:
+  - mix do deps.get, compile --warnings-as-errors, coveralls.json
+after_success:
+  - bash <(curl -s https://codecov.io/bash)
+# See: github.com/dwyl/repo-badges#documentation
+after_script:
+  - MIX_ENV=docs mix deps.get
+  - MIX_ENV=docs mix inch.report
```
Comment (shouston3) [inline .travis.yml]: `MIX_ENV=docs` is not a typical `MIX_ENV`. As far as I can see we are also doing nothing in that `MIX_ENV`, so it behaves just as `MIX_ENV=dev` would behave.
I was intregued as to the use case for `MIX_ENV=docs`, so doing a search for `MIX_ENV=docs` in github, I found that it is used in places like here: https://github.com/zamith/tomlex/blob/master/.travis.yml#L9
But that repo makes use of `MIX_ENV=docs` in the `mix.exs` file in that only the deps which are used for documentation are compiled. We are not doing that in our `mix.exs` file, so we should either remove `MIX_ENV=docs` or update our mix file to be similar to https://github.com/zamith/tomlex/blob/master/mix.exs

File: assets/js/socket.js (hunk 7 above)
Comment (shouston3) [reply assets/js/socket.js]: Nope, this is just what comes with: `mix phx.new ...`. This entire file is not run.

File: config/config.exs [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  pubsub: [name: App.PubSub,
+           adapter: Phoenix.PubSub.PG2]
+
+# Configures Elixir's Logger
+config :logger, :console,
+  format: "$time $metadata[$level] $message\n",
+  metadata: [:request_id]
+
+# Configures pre-commit commands
+config :pre_commit, commands: ["credo --strict", "test"], verbose: true
```
Comment (shouston3) [inline config/config.exs]: Oh, this is cool 👍

File: assets/js/socket.js (hunk 7 above)
Comment (shouston3) [reply assets/js/socket.js]: Nope, this is just what comes with: `mix phx.new ...`. This entire file is not run.

File: .travis.yml (hunk 9 above)
Comment (shouston3) [reply .travis.yml]: We could actually consider not using credo if we use `mix format` instead offered in (elixir 1.6)

File: .travis.yml (hunk 1 above)
Comment (shouston3) [reply .travis.yml]: Actually, this should be left on. It does not error for third party warnings, see: https://github.com/dwyl/hits-elixir/issues/19#issuecomment-372829963

File: .travis.yml (hunk 9 above)
Comment (SimonLab) [reply .travis.yml]: Is 1.6 available on Travis, I'm not sure :thinking: ?

Comment (shouston3) [reply .travis.yml]: https://github.com/phoenixframework/phoenix/blob/master/.travis.yml#L5

Comment (SimonLab) [reply .travis.yml]: I think the reason why we don't have 1.6 is that the version on Heroku can be up to 1.5:
https://github.com/HashNuke/heroku-buildpack-elixir/blob/master/elixir_buildpack.config
So I would prefer to keep the same version between Heroku and Travis to make sure the tests and the app are run on the same version.
But at the end I think the version has been updated on the project since:
https://github.com/dwyl/github-backup/blob/7c7c2de83a6f2047925efa0da5ffd86dcaa855f2/.travis.yml#L3
